ports/geckolib/gecko_bindings/sugar/refptr.rs

```diff
@@ -0,0 +1,113 @@
+use std::marker::PhantomData;
+use std::mem::{forget, transmute};
+use std::ptr;
+use std::sync::Arc;
+
+/// Helper trait for conversions between FFI Strong/Borrowed types and Arcs
+///
+/// Should be implemented by types which are passed over FFI as Arcs
+/// via Strong and Borrowed
+pub unsafe trait HasArcFFI where Self: Sized {
+    /// Gecko's name for the type
+    /// This is equivalent to ArcInner<Self>
+    type FFIType: Sized;
+
+    /// Given a non-null borrowed FFI reference, this produces a temporary
+    /// Arc which is borrowed by the given closure and used.
+    /// Panics on null.
+    fn with<F, Output>(raw: Borrowed<Self::FFIType>, cb: F) -> Output
+               where F: FnOnce(&Arc<Self>) -> Output {
+        Self::maybe_with(raw, |opt| cb(opt.unwrap()))
+    }
+
+    /// Given a maybe-null borrowed FFI reference, this produces a temporary
+    /// Option<Arc> (None if null) which is borrowed by the given closure and used
+    fn maybe_with<F, Output>(maybe_raw: Borrowed<Self::FFIType>, cb: F) -> Output
+                         where F: FnOnce(Option<&Arc<Self>>) -> Output {
+        cb(Self::borrowed_as(&maybe_raw))
+    }
+
+    /// Given a non-null strong FFI reference, converts it into an Arc.
+    /// Panics on null.
+    fn into(ptr: Strong<Self::FFIType>) -> Arc<Self> {
+        assert!(!ptr.is_null());
+        unsafe { transmute(ptr) }
+    }
+
+    fn borrowed_as<'a>(ptr: &'a Borrowed<'a, Self::FFIType>) -> Option<&'a Arc<Self>> {
+        unsafe {
+            if ptr.is_null() {
+                None
+            } else {
+                Some(transmute::<&Borrowed<_>, &Arc<_>>(ptr))
+            }
+        }
+    }
+
+    /// Converts an Arc into a strong FFI reference.
+    fn from_arc(owned: Arc<Self>) -> Strong<Self::FFIType> {
+        unsafe { transmute(owned) }
+    }
+
+    /// Artificially increments the refcount of a borrowed Arc over FFI.
+    unsafe fn addref(ptr: Borrowed<Self::FFIType>) {
```

Comment:
I find it odd that addref panics if passed a null reference, but release doesn't. Perhaps try to be more consistent.